Boost: Add notice if legacy concatenation files are used

donnchawp opened the pull request:
Fixes [41675](https://github.com/Automattic/jetpack/issues/41675).

If the new static cache files introduced in #41056 are not being used by a server then notify the user so they can fix the problem.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Add a notice in the CSS and JS concatenation modules. (duplication? Maybe it should be a single notice?)
* Add a useStaticMinification hook to fetch the static cache files status
* Register a new readonly status to read the static_minification site option as a bool.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
pc9hqz-3kc-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Apply PR and `jetpack rebuild plugins/boost` to build everything.
* Make sure one or both Concatenation modules are enabled. 
 * If it's using the new static files you won't see a notice.
 * To enforce it using the legacy delivery method:
   * Create wp-content/.htaccess with `RewriteEngine Off` in it.
   * Toggle off both Concatenation modules, and toggle them back on.
* A notice should appear in the JS module if they are both enabled.
![Screenshot 2025-02-14 at 14 25 54](https://github.com/user-attachments/assets/aca18af3-f49a-4c4e-be45-2ca884b60ddf)
* Disable the Concatenate JS module, the notice should now appear under Concatenate CSS.
* Enable the Concatenate JS module, the notice should now appear under Concatenate JS.
* Define a constant `JETPACK_BOOST_DISABLE_404_TESTER` within your `wp-config.php`
* Refresh the page. The notice should no longer be visible under any notice.
* Comment out or rename `.htaccess` to `htaccess` added previously.
* Disable both Concatenate modules, and re-enable. The notice should now be gone.
* Dismiss the notice and refresh the page, the notice should now be dismissed permanently.

donnchawp committed:
Add notice if legacy concatenation files are used

donnchawp committed:
changelog

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into update/boost/add-concatenation-notice

donnchawp committed:
Only show contactenation notice once, and show hide buttons

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into update/boost/add-concatenation-notice

LiamSarsfield committed:
Boost: Improve Static Minification and Module Handling

- Refactor `useStaticMinification` hook to return more detailed query data
- Update Minification modules to use new hook and add immediate 404 tester on activation
- Modify module UI to conditionally show notices based on loading and enabled states
- Add option to run 404 tester immediately during module activation

donnchawp committed:
Make it read write instead of readonly

LiamSarsfield committed:
Boost: Refactor Minification Modules and Static Minification Notice

- Remove Minify_Notice_Entry and replace with a readonly option for static minification
- Add new components for Minify CSS and JS modules
- Introduce a LegacyMinifyNotice component for displaying static minification warning
- Update performance history hooks to support static minification notice
- Simplify index page by extracting minification modules into separate components

LiamSarsfield committed:
Boost: Refactor Minify Legacy Notice and Related Components

- Rename and update MinifyLegacyNotice component with more detailed information
- Update data sync keys and hooks for legacy minify notice
- Modify minify CSS and JS modules to use new legacy notice logic
- Remove static-minification store and replace with minify store
- Add support for dismissing legacy minify notice with analytics tracking

LiamSarsfield committed:
Boost: Minor Improvements to Data Sync and 404 Tester

- Add namespace definition for Jetpack Boost data sync
- Rename parameter in jetpack_boost_404_setup for clarity

haqadn commented on projects/plugins/boost/app/lib/minify/functions-helpers.php:
```diff
@@ -350,12 +356,12 @@ function jetpack_boost_minify_serve_concatenated() {
  * @return void
  */
-function jetpack_boost_minify_activation() {
+function jetpack_boost_minify_activation( $run_tester_immediately = false ) {
```
Are there situations where we do not need to run 404 tester immediately on minify activation? If not, I think we should pass a boolean directly to the `jetpack_boost_404_setup` function.

LiamSarsfield replied:
We do run the 404 tester activation if we detect the plugin has been updated, and they have either Minify Module enabled. 
https://github.com/Automattic/jetpack/blob/55c6ab89bf43cb9e8912963c83091507f6a76de8/projects/plugins/boost/app/class-jetpack-boost.php#L152-L167

Though I don't think there would be an issue with running the 404 tester itself immediately in this scenario either. So I'm updating the code to apply your feedback.

haqadn commented on projects/plugins/boost/app/lib/minify/functions-helpers.php:
```diff
@@ -188,9 +188,15 @@ function jetpack_boost_page_optimize_remove_concat_base_prefix( $original_fs_pat
  * Schedule a cronjob for the 404 tester, if one isn't already scheduled.
  */
-function jetpack_boost_page_optimize_schedule_404_tester() {
+function jetpack_boost_page_optimize_schedule_404_tester( $run_immediately = false ) {
```
Perhaps we can skip the chain of boolean and run immediately always if no cron is scheduled?

LiamSarsfield replied:
Good point, I've updated the PR to remove the chain of booleans, and in relation to [this comment](https://github.com/Automattic/jetpack/pull/41604#discussion_r1957900022).

haqadn commented on projects/plugins/boost/wp-js-data-sync.php:
```diff
@@ -259,6 +259,19 @@ function () {
+			return false;
+		}
+
+		// Otherwise show it if the 404 tester determined it can't be used.
+		return ! (bool) get_site_option( 'jetpack_boost_static_minification' );
```
The flashing notice issue mentioned above is probably because of this. Before the 404 tester has a chance to run, the default value is false as the option doesn't exist yet. 

As a result, when you enable a module, initially the notice will show. After it has a chance to fetch, it hides the notice again. I think something like the following might work:

```suggestion
		/*
		 * Until the 404 tester running, the option will not be created and it will return false.
		 * However, we don't want to show the notice before the tester has a chance to run.
		 * We only want to show the notice when the tester has already run and determined that
		 * it will use the fallback system.
		 */
		return get_site_option( 'jetpack_boost_static_minification' ) !== '';
```

LiamSarsfield replied:
Nice catch, I've updated the PR with something similar to your solution.

haqadn commented on projects/plugins/boost/app/assets/src/js/features/minify-legacy-notice/minify-legacy-notice.tsx:
```diff
@@ -0,0 +1,45 @@
+			>
+				<p>
+					{ __(
+						'You are using the legacy cache delivery method for concatenated files.',
```
I am a bit unsure about this. But, maybe we don't want to call it legacy in a user-facing way? I think the notice should instead say something like: "You can improve the speed of concatenated files and reduce the load on WordPress." ... Learn More.

LiamSarsfield replied:
You're right legacy may not be the correct word here, we also don't mention it's legacy in [the documentation](https://jetpack.com/support/jetpack-boost/troubleshooting-concatenated-css-or-javascript-delivery-methods/). I've updated the PR with that text, and Oxford comma.
 ![Screenshot 2025-02-17 at 12 14 20](https://github.com/user-attachments/assets/1d62bb47-03b1-4384-9925-fd3d0cf313bb)

haqadn reviewed (changes requested):
Overall looks good. I have a couple of nitpicks and noticed an issue.

The notice flashes for a moment when you enable any of the concat modules for the first time.

LiamSarsfield committed:
Boost: Refactor Minify Activation and Legacy Notice Logic

- Remove immediate 404 tester parameter from activation functions
- Update legacy notice to conditionally display based on static minification state
- Simplify 404 tester scheduling logic
- Modify legacy notice text to focus on performance improvement

LiamSarsfield committed:
Remove unnecessary argument

haqadn commented on projects/plugins/boost/app/assets/src/js/lib/stores/minify.ts:
```diff
@@ -28,3 +28,9 @@ export const useMetaQuery = ( key: MinifyMetaKeys, onSuccess?: ( newState: strin
 };
+
+export const useShowMinifyLegacy = () => {
+	const [ query ] = useDataSync( 'jetpack_boost_ds', 'minify_legacy_notice', z.boolean() );
+
+	return query;
```
It's a nitpick but since we aren't using other properties, can we narrow it down to data? This will make it cleaner on the end it is being used.

```suggestion
	const [ query: { data } ] = useDataSync( 'jetpack_boost_ds', 'minify_legacy_notice', z.boolean() );

	return data;
```

LiamSarsfield replied:
We're using the `refetch` function within the Minify components so the query object is needed unfortunately, e.g.
https://github.com/Automattic/jetpack/blob/866d7dd70efba716d1ae55bb5f23793f7dbc2cd6/projects/plugins/boost/app/assets/src/js/features/minify-js/minify-js.tsx#L22

haqadn reviewed (approved):
I think it's ready to merge. But, I have one tiny suggestion.

LiamSarsfield committed:
Merge branch 'trunk' into update/boost/add-concatenation-notice
